File: app/src/main/java/com/coinninja/coinkeeper/service/runner/GetIncomingInviteRunner.java
package com.coinninja.coinkeeper.service.runner;

import android.util.Log;

import com.coinninja.coinkeeper.model.helpers.InviteTransactionSummaryHelper;
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient;
import com.coinninja.coinkeeper.service.client.model.ReceivedInvite;

import java.io.IOException;
import java.util.List;

import javax.inject.Inject;

import retrofit2.Response;

public class GetIncomingInviteRunner implements Runnable {
    private static final String TAG = InviteContactRunner.class.getSimpleName();

    private final SignedCoinKeeperApiClient client;
    private final InviteTransactionSummaryHelper inviteTransactionSummaryHelper;

    @Inject
    public GetIncomingInviteRunner(SignedCoinKeeperApiClient client,
                                   InviteTransactionSummaryHelper inviteTransactionSummaryHelper) {
        this.client = client;
        this.inviteTransactionSummaryHelper = inviteTransactionSummaryHelper;
    }

    @Override
    public void run() {
        Response response = client.getReceivedInvites();
        if (response.isSuccessful()) {
            writeInvitesToDatabase((List<ReceivedInvite>) response.body());
        } else {
            logError(response);
        }
    }

    private void writeInvitesToDatabase(List<ReceivedInvite> receivedInvites) {
        for (ReceivedInvite invite : receivedInvites) {
            inviteTransactionSummaryHelper.saveReceivedInviteTransaction(invite);
        }
    }

    private void logError(Response response) {
        Log.d(TAG, "|---- Received Invite failed");
        Log.d(TAG, "|------ statusCode: " + String.valueOf(response.code()));
        try {
            Log.d(TAG, "|--------- message: " + response.errorBody().string());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: app/src/main/java/com/coinninja/coinkeeper/service/runner/GetIncomingInviteRunner.kt
package com.coinninja.coinkeeper.service.runner

import app.coinninja.cn.thunderdome.repository.ThunderDomeRepository
import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.model.db.enums.Type
import com.coinninja.coinkeeper.model.helpers.InviteTransactionSummaryHelper
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient
import com.coinninja.coinkeeper.service.client.model.ReceivedInvite
import com.coinninja.coinkeeper.util.CNLogger
import javax.inject.Inject

@Mockable
class GetIncomingInviteRunner @Inject constructor(
        internal val client: SignedCoinKeeperApiClient,
        internal val inviteTransactionSummaryHelper: InviteTransactionSummaryHelper,
        internal val thunderDomeRepository: ThunderDomeRepository,
        internal val logger: CNLogger
) : Runnable {

    @Suppress("UNCHECKED_CAST")
    override fun run() {
        val response = client.receivedInvites
        if (response.isSuccessful) {
            (response.body() as List<ReceivedInvite>?)?.let {
                writeInvitesToDatabase(it)
            }
        } else {
            logger.logError(this::class.java.simpleName, "|---- Received Invite failed", response)
        }
    }

    internal fun writeInvitesToDatabase(receivedInvites: List<ReceivedInvite>) {
        for (invite in receivedInvites) {
            saveInvite(invite)
        }
    }

    internal fun saveInvite(invite: ReceivedInvite) {
        val invite = inviteTransactionSummaryHelper.saveReceivedInviteTransaction(invite)
        if (invite.type == Type.LIGHTNING_RECEIVED) {
            thunderDomeRepository.createSettlementForInvite(invite.id, invite.toUser.id, invite.fromUser.id, invite.sentDate)
        }
    }

}
